### PR TITLE
Making the registry iterable. See #1918.

### PR DESCRIPTION
**Pull request by joeldenning**
Example usage:

```js
for (let entry of System) {
  console.log(entry) // ['http://hi', {firstExport: 'value'}]
}
```

### CODE REVIEW

File: src/features/registry.js [hunk 1]
```diff
@@ -58,4 +58,20 @@ systemJSPrototype.delete = function (id) {
         depLoad.i.splice(importerIndex, 1);
     });
   return delete this[REGISTRY][id];
-};
\ No newline at end of file
+};
+
+systemJSPrototype[Symbol && Symbol.iterator ? Symbol.iterator : '@@iterator'] = function() {
```
Comment (guybedford) [inline src/features/registry.js]: Note the `Symbol` check is missing a `typeof Symbol` here.

I would prefer we only set the `Symbol.iterator` property when `Symbol` is defined, and we likely don't even need to do the `Symbol.iterator` check as I believe all browsers supporting symbols will support `Symbol.iterator`?

File: docs/api.md [hunk 2]
```diff
@@ -87,3 +87,14 @@ System.set('http://site.com/normalized/module/name.js', {
 `module` is an object of names to set as the named exports.
 
 If `module` is an existing Module Namespace, it will be used by reference.
+
+#### System[Symbol.iterator]() -> Iterator
```
Comment (guybedford) [inline docs/api.md]: Can we make this `System.entries()` rather, and then say "also available at `System[Symbol.iterator]`.

File: src/features/registry.js [hunk 3]
```diff
@@ -58,4 +58,20 @@ systemJSPrototype.delete = function (id) {
...
+  return {
+    next() {
+      if (index === keys.length) {
+        return {done: true};
+      } else {
+        const key = keys[index++];
+        return {done: false, value: [key, registry[key]]};
+      }
+    }
+  }
```
Comment (guybedford) [inline src/features/registry.js]: Semicolon.

File: src/features/registry.js [hunk 4]
```diff
@@ -58,4 +58,20 @@ systemJSPrototype.delete = function (id) {
...
-};
\ No newline at end of file
+};
+
+systemJSPrototype[Symbol && Symbol.iterator ? Symbol.iterator : '@@iterator'] = function() {
+  const registry = this[REGISTRY]
+  const keys = Object.keys(registry);
+  let index = 0;
+  return {
+    next() {
```
Comment (guybedford) [inline src/features/registry.js]: One way to write this a little more succinctly might be to skip the branch and just have (again, byte shaving):

```js
const key = keys[index++];
return { done: index > keys.length, value: key && [key, registry[key]] };
```

**Comment by joeldenning** [comment]
I just pushed the changes you suggested, and also found a bug where i was forgetting to do `registry[key].n` instead of just `registry[key]`

File: src/features/registry.js [hunk 5]
```diff
@@ -58,4 +58,25 @@ systemJSPrototype.delete = function (id) {
...
-};
\ No newline at end of file
+};
+
+systemJSPrototype.entries = function () {
+  const registry = this[REGISTRY];
+  return Object.keys(registry).map(function (key) {
+    return [key, registry[key].n];
+  });
+}
```
Comment (guybedford) [inline src/features/registry.js]: I was thinking more like the `Map.prototype.entries()` approach which is just the same as the iterator.

File: src/features/registry.js [hunk 6]
```diff
@@ -58,4 +58,25 @@ systemJSPrototype.delete = function (id) {
...
+
+if (typeof Symbol !== 'undefined') {
+  systemJSPrototype[Symbol.iterator] = function () {
```
Comment (guybedford) [inline src/features/registry.js]:
```suggestion
  systemJSPrototype[Symbol.iterator] = systemJSPrototype.entries;
```

**Comment by guybedford** [comment]
I've gone ahead and merged - will adjust the iterator separately.

**Comment by guybedford** [comment]
Updated in https://github.com/systemjs/systemjs/commit/532fdbddede7c2b7664f4897e31da742e7e1094c.

**Pull request merged**